### ISSUE TITLE
Pod stuck pending with volume node affinity conflict

### ISSUE THREAD

**Issue by ben-nelson-nbcuni** [aws/karpenter#2971]
### Version

__Karpenter Version:__ v0.19.2

__Kubernetes Version:__ v1.21.14

### Expected Behavior

I have an EKS cluster with 3 subnets split across 3 zones. I have a statefulset that has the following node affinity:
```
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: eks.amazonaws.com/capacityType
            operator: In
            values:
            - ON_DEMAND
        - matchExpressions:
          - key: karpenter.sh/capacity-type
            operator: In
            values:
            - on-demand
```
When a pod in that statefulset is pending, Karpenter should spin up a new node in the zone that matches the statefulset's volume even though the zone isn't directly specified in the pod's node affinity.

### Actual Behavior

The pod is stuck in pending. The kubernetes logs have the following 2 lines:
```
Warning  FailedScheduling  2m28s (x8 over 9m3s)  default-scheduler  0/6 nodes are available: 1 node(s) had volume node affinity conflict, 2 node(s) didn't match Pod's node affinity/selector, 3 Insufficient cpu.                              │
Normal   Nominated         18s (x6 over 10m)     karpenter          Pod should schedule on ip-10-20-93-221.us-east-2.compute.internal 
```
It's worth noting that `ip-10-20-93-221.us-east-2.compute.internal ` is in us-east-2b, but the Persistent Volume for the statefulset is in us-east-2a.

When I view the Karpenter logs, I see the following:
```
2022-12-01T22:13:17.380Z    DEBUG    controller.provisioning    relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution[0]={"matchExpressions":[{"key":"eks.amazonaws.com/capacityType","operator":"In","values":["ON_DEMAND"]},{"key":"topology.kubernetes.io/zone","operator":"In","values":["us-east-2a"]},{"key":"topology.kubernetes.io/region","operator":"In","values":["us-east-2"]}]}    {"commit": "470aa83", "pod": "concourse/concourse-worker-1"}
```

According to the logs, Karpenter appears to know that the zone needs to be `us-east-2a`, but I'm guessing that it's only attaching that affinity to the first of the 2 possible node affinities (we have 2, one for nodes from EKS ASGs and one for nodes from Karpenter).

### Steps to Reproduce the Problem

1. Add 2 node affinities to a statefulset. The second one should match Karpenter's annotation requirement.
2. Scale up the statefulset to a high number to get the Persistent Volumes allocated to random zones.
3. Scale down the statefulset to remove the pods and wait for Karpenter to scale down the nodes.
4. Scale the statefulsets back up. Some pods will be stuck in pending as Karpenter has issues attaching the volume node affinity to all of the pod node affinities options.

### Resource Specs and Logs

[karpenter.log](https://github.com/aws/karpenter/files/10136575/karpenter.log)

Specs:
```
---
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  annotations:
    meta.helm.sh/release-name: karpenter-config-1-21
    meta.helm.sh/release-namespace: kube-system
  creationTimestamp: "2022-12-01T20:33:48Z"
  generation: 2
  labels:
    app.kubernetes.io/instance: karpenter-config-1-21
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: karpenter-config
    helm.sh/chart: karpenter-config-1.0.0
  name: ((redacted))
  resourceVersion: "38994834"
  uid: 6d79390e-3aff-41fb-80e6-339dd57bd75b
spec:
  providerRef:
    name: ((redacted))
  requirements:
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - spot
    - on-demand
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
  ttlSecondsAfterEmpty: 30
  weight: 100
status:
  resources:
    attachable-volumes-aws-ebs: "75"
    cpu: "8"
    ephemeral-storage: 62877636Ki
    memory: 23613Mi
    pods: "92"
---
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
metadata:
  annotations:
    meta.helm.sh/release-name: karpenter-config-1-21
    meta.helm.sh/release-namespace: kube-system
  creationTimestamp: "2022-12-01T20:33:47Z"
  generation: 1
  labels:
    app.kubernetes.io/instance: karpenter-config-1-21
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: karpenter-config
    helm.sh/chart: karpenter-config-1.0.0
  name: ((redacted))
  resourceVersion: "38964472"
  uid: 01029fc9-70c1-4485-87c1-810fad34ffc9
spec:
  amiSelector:
    eks/version: "1.21"
  securityGroupSelector:
    karpenter.sh/discovery: ((redacted))
  subnetSelector:
    karpenter.sh/discovery: ((redacted))
  tags:
    aws-node-termination-handler/managed: ""
    inv_cluster_id: ((redacted))
    inv_cluster_type: kubernetes
    inv_cost_centre: N/A
    inv_environment_id: lab
    inv_node_type: worker
    inv_node_version: "1.21"
    inv_proposition_id: nbcu
    inv_resident_id: N/A
---
apiVersion: v1
kind: Pod
metadata:
  annotations:
    checksum/secrets: 88c4b5ebe42d280bdf6c6596350e64ad9792db58541a53bf6b262fe791f9035c
    kubernetes.io/psp: eks.privileged
  creationTimestamp: "2022-12-01T21:29:04Z"
  generateName: concourse-worker-
  labels:
    app: concourse-worker
    controller-revision-hash: concourse-worker-fc68d788d
    release: concourse
    statefulset.kubernetes.io/pod-name: concourse-worker-1
  name: concourse-worker-1
  namespace: concourse
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: StatefulSet
    name: concourse-worker
    uid: 80b8f2b0-58d3-4bb5-b296-573443ff7cd3
  resourceVersion: "38994886"
  uid: 6c1bd790-78b1-4977-915a-5120d7b2ea77
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: eks.amazonaws.com/capacityType
            operator: In
            values:
            - ON_DEMAND
        - matchExpressions:
          - key: karpenter.sh/capacity-type
            operator: In
            values:
            - on-demand
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - podAffinityTerm:
          labelSelector:
            matchLabels:
              app: concourse-worker
              release: concourse
          topologyKey: kubernetes.io/hostname
        weight: 100
  containers:
  - args:
    - worker
    env:
    - name: CONCOURSE_SWEEP_INTERVAL
      value: 30s
    - name: CONCOURSE_CONNECTION_DRAIN_TIMEOUT
      value: 1h
    - name: CONCOURSE_HEALTHCHECK_BIND_IP
      value: 0.0.0.0
    - name: CONCOURSE_HEALTHCHECK_BIND_PORT
      value: "8888"
    - name: CONCOURSE_HEALTHCHECK_TIMEOUT
      value: 5s
    - name: CONCOURSE_DEBUG_BIND_IP
      value: 127.0.0.1
    - name: CONCOURSE_DEBUG_BIND_PORT
      value: "7776"
    - name: CONCOURSE_WORK_DIR
      value: /concourse-work-dir
    - name: CONCOURSE_BIND_IP
      value: 127.0.0.1
    - name: CONCOURSE_BIND_PORT
      value: "7777"
    - name: CONCOURSE_LOG_LEVEL
      value: info
    - name: CONCOURSE_TSA_HOST
      value: concourse-web-worker-gateway:2222
    - name: CONCOURSE_TSA_PUBLIC_KEY
      value: /concourse-keys/host_key.pub
    - name: CONCOURSE_TSA_WORKER_PRIVATE_KEY
      value: /concourse-keys/worker_key
    - name: CONCOURSE_BAGGAGECLAIM_LOG_LEVEL
      value: info
    - name: CONCOURSE_BAGGAGECLAIM_BIND_IP
      value: 127.0.0.1
    - name: CONCOURSE_BAGGAGECLAIM_BIND_PORT
      value: "7788"
    - name: CONCOURSE_BAGGAGECLAIM_DEBUG_BIND_IP
      value: 127.0.0.1
    - name: CONCOURSE_BAGGAGECLAIM_DEBUG_BIND_PORT
      value: "7787"
    - name: CONCOURSE_BAGGAGECLAIM_BTRFS_BIN
      value: btrfs
    - name: CONCOURSE_BAGGAGECLAIM_MKFS_BIN
      value: mkfs.btrfs
    - name: CONCOURSE_VOLUME_SWEEPER_MAX_IN_FLIGHT
      value: "5"
    - name: CONCOURSE_CONTAINER_SWEEPER_MAX_IN_FLIGHT
      value: "5"
    image: concourse/concourse:7.8.3
    imagePullPolicy: IfNotPresent
    lifecycle:
      preStop:
        exec:
          command:
          - /bin/bash
          - /pre-stop-hook.sh
    livenessProbe:
      failureThreshold: 5
      httpGet:
        path: /
        port: worker-hc
        scheme: HTTP
      initialDelaySeconds: 10
      periodSeconds: 15
      successThreshold: 1
      timeoutSeconds: 3
    name: concourse-worker
    ports:
    - containerPort: 8888
      name: worker-hc
      protocol: TCP
    resources:
      limits:
        cpu: "2"
        memory: 4Gi
      requests:
        cpu: "2"
        memory: 4Gi
    securityContext:
      privileged: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /concourse-keys
      name: concourse-keys
      readOnly: true
    - mountPath: /concourse-work-dir
      name: concourse-work-dir
    - mountPath: /pre-stop-hook.sh
      name: pre-stop-hook
      subPath: pre-stop-hook.sh
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-knn29
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  hostname: concourse-worker-1
  initContainers:
  - args:
    - -ce
    - |-
      for v in $((btrfs subvolume list --sort=-ogen "/concourse-work-dir" || true) | awk '{print $9}'); do
        (btrfs subvolume show "/concourse-work-dir/$v" && btrfs subvolume delete "/concourse-work-dir/$v") || true
      done
      rm -rf "/concourse-work-dir/*"
    command:
    - /bin/bash
    image: concourse/concourse:7.8.3
    imagePullPolicy: IfNotPresent
    name: concourse-worker-init-rm
    resources: {}
    securityContext:
      privileged: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /concourse-work-dir
      name: concourse-work-dir
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-knn29
      readOnly: true
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  priorityClassName: default
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: concourse-worker
  serviceAccountName: concourse-worker
  subdomain: concourse-worker
  terminationGracePeriodSeconds: 60
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: concourse-work-dir
    persistentVolumeClaim:
      claimName: concourse-work-dir-concourse-worker-1
  - configMap:
      defaultMode: 420
      name: concourse-worker
    name: pre-stop-hook
  - name: concourse-keys
    secret:
      defaultMode: 256
      items:
      - key: host-key-pub
        path: host_key.pub
      - key: worker-key
        path: worker_key
      secretName: concourse-worker
  - name: kube-api-access-knn29
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2022-12-01T21:29:04Z"
    message: '0/6 nodes are available: 1 node(s) had volume node affinity conflict,
      2 node(s) didn''t match Pod''s node affinity/selector, 3 Insufficient cpu.'
    reason: Unschedulable
    status: "False"
    type: PodScheduled
  phase: Pending
  qosClass: Burstable
---
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubernetes.io/createdby: aws-ebs-dynamic-provisioner
    pv.kubernetes.io/bound-by-controller: "yes"
    pv.kubernetes.io/provisioned-by: kubernetes.io/aws-ebs
  creationTimestamp: "2022-11-16T23:10:32Z"
  finalizers:
  - kubernetes.io/pv-protection
  labels:
    topology.kubernetes.io/region: us-east-2
    topology.kubernetes.io/zone: us-east-2a
  name: pvc-746adf38-9a7f-451b-a1c8-a406b6c6afe5
  resourceVersion: "32647356"
  uid: edc184b0-2ab1-4054-8f0b-4c1028b912a9
spec:
  accessModes:
  - ReadWriteOnce
  awsElasticBlockStore:
    fsType: ext4
    volumeID: aws://us-east-2a/vol-0bbb63e3750d7e826
  capacity:
    storage: 80Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: concourse-work-dir-concourse-worker-1
    namespace: concourse
    resourceVersion: "32647309"
    uid: 746adf38-9a7f-451b-a1c8-a406b6c6afe5
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: topology.kubernetes.io/zone
          operator: In
          values:
          - us-east-2a
        - key: topology.kubernetes.io/region
          operator: In
          values:
          - us-east-2
  persistentVolumeReclaimPolicy: Delete
  storageClassName: gp2
  volumeMode: Filesystem
status:
  phase: Bound
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by tzneal** [comment]
Sorry, this didn't make v0.20.0, but should be released soon.

**Comment by bnhoogvliet** [comment]
Is this included in release [v0.21.0](https://github.com/aws/karpenter/releases/tag/v0.21.0) ? 
Because I don't see it mentioned.
